core/modules/loader/appClientNum.cc

```diff
@@ -0,0 +1,262 @@
+            }
+            iter = kList.erase(iter);
+        } else {
+            ++iter;
+        }
+    }
+    return kList.empty();
+}
+
+
+KeyInfoData::Ptr clientAdd(CentralClient& central, uint64_t j) {
+    CompositeKey cKey(j);
+    int chunk = j%10000;
+    int subchunk = j%100;
+    return central.keyInsertReq(cKey, chunk, subchunk);
+}
+
+
+/// @return true if the list is empty.
+bool keyLookupListClean(KeyInfoDataList& kList, int& successCount, int& failedCount) {
+    for(auto iter=kList.begin(); iter != kList.end();) {
+        if (*iter == nullptr || (*iter)->isFinished()) {
+            KeyInfoData::Ptr const& kPtr = *iter;
+            if (kPtr->success) {
+                // check the values
+                uint64_t j = kPtr->key.kInt;
+                // expected chunk and subchunk values.
+                int expChunk = j%10000;
+                int expSubchunk = j%100;
+                if (kPtr->chunk == expChunk && kPtr->subchunk == expSubchunk) {
+                    ++successCount;
+                } else {
+                    ++failedCount;
+                    LOGS(_log, LOG_LVL_WARN, "lookup failed bad values expected c=" << expChunk <<
+                            " sc=" << expSubchunk << " found=" << *kPtr);
+                }
+            } else {
+                ++failedCount;
+                LOGS(_log, LOG_LVL_WARN, "lookup failed " << *kPtr);
+            }
+            iter = kList.erase(iter);
+        } else {
+            ++iter;
+        }
+    }
+    return kList.empty();
+}
+
+
+KeyInfoData::Ptr clientAddLookup(CentralClient& central, uint64_t j) {
+    CompositeKey cKey(j);
+    return central.keyLookReq(cKey);
+}
+
+
+int main(int argc, char* argv[]) {
+    std::string cCfgFile("core/modules/loader/config/client1.cnf");
+    if (argc < 3) {
+        LOGS(_log, LOG_LVL_ERROR, "usage: appClientNum <startingNumber> <endingNumber> <optional config file name>");
+        exit(-1);
+    }
+    uint64_t numStart = std::stoi(argv[1]);
+    uint64_t numEnd   = std::stoi(argv[2]);
+    if (argc > 3) {
+        cCfgFile = argv[3];
+    }
+    LOGS(_log, LOG_LVL_INFO, "start=" << numStart << " end=" << numEnd << " cCfg=" << cCfgFile);
+
+    std::string ourHost;
+    {
+        char hName[300];
+        if (gethostname(hName, sizeof(hName)) < 0) {
+            LOGS(_log, LOG_LVL_ERROR, "Failed to get host name errno=" << errno);
+            exit(-1);
+        }
+        ourHost = hName;
+    }
+
+    boost::asio::io_service ioService;
+
+    ClientConfig cCfg(cCfgFile);
+    CentralClient cClient(ioService, ourHost, cCfg);
+    try {
+        cClient.start();
+    } catch (boost::system::system_error const& e) {
+        LOGS(_log, LOG_LVL_ERROR, "cWorker.start() failed e=" << e.what());
+        exit(-1);
+    }
+
+    // Need to start several threads so messages aren't dropped while being processed.
+    cClient.run();
+    cClient.run();
+    cClient.run();
+    cClient.run();
```

Comment:
I guess each creates a new BOOST ASIO thread. This might be better be done explicitly in a loop:
```
int const numClients = 4;
for (int i=0; i<numClients; ++i) {
    cClient.run();
}
```